stripped/colabs/agregacje_kilkuletnie_stripped.py

```diff
@@ -0,0 +1,669 @@
+        .master("local")\
+        .appName("Colab")\
+        .config('spark.ui.port', '4050')\
+        .getOrCreate()
+
+# Commented out IPython magic to ensure Python compatibility.
+# %%time
+# nasa = default_loader(spark)
+# nasa.show(5)
+
+nasa.createOrReplaceTempView("nasa")
+
+"""# Średnie wieloletnie dla poszczególnych miesięcy"""
+
+def overYearMonthStats(columns: list, n: int):
+    """
+    Funkcja licząca średnie dla wybranych zmiennych dla danych współrzędnych geograficznych
+    dla każdego z 12 miesięcy od roku w rozważanym pomiarze do n lat wstecz (tj. gdyby n=3, to
+    średnia zostałaby wyliczona z rozważanego roku oraz 3 lat przed nim, czyli w praktyce mielibyśmy
+    średnią z łącznie 4 lat).
+
+    :param columns: lista nazw rozważanych kolumn ze zbioru danych
+    :param n: liczba lat wstecz, z których chcemy wyliczyć średnią
+    """
+
+    if not columns:
+      raise ValueError("Podana musi zostać co najmniej jedna zmienna")
+
+    windowSpec = Window.partitionBy("lon", "lat", "Month").orderBy("Year").rowsBetween(-n, 0)
+
+    nasa_window = nasa
+    for column in columns:
+        average_column_name = "monthly_avg_" + str(n) + "years_" + column
+        nasa_window = nasa_window.withColumn(average_column_name, F.avg(F.col(column)).over(windowSpec))
+
+    return nasa_window
+
+"""Zastosujmy teraz powyższą funkcję na naszym datasecie, dla każdej z rozważanych kolumn pomiarowych, przyjmując n=3:"""
+
+columns_to_avg = ["Rainf", "Evap", "AvgSurfT", "Albedo", "SoilT_40_100cm", "GVEG", "PotEvap", "RootMoist", "SoilM_100_200cm"]
+
+result_df_3 = overYearMonthStats(columns_to_avg, n=3)
+result_df_3.show(15)
+
+"""Wyodrębnijmy teraz kilka lat do oddzielnych dataframe'ów i zapiszmy je:"""
+
+# 2023
+nasa_2023_monthly_avg_3 = result_df_3.filter(f'Year = {2023}')
+pd_2023_monthly_avg_3 = nasa_2023_monthly_avg_3.toPandas()
+
+pd_2023_monthly_avg_3.to_csv('/content/drive/MyDrive/BigMess/NASA/monthly_avg/3years/NASA_monthly_avg_2023.csv')
+
+# 2022
+nasa_2022_monthly_avg_3 = result_df_3.filter(f'Year = {2022}')
+pd_2022_monthly_avg_3 = nasa_2022_monthly_avg_3.toPandas()
+
+pd_2022_monthly_avg_3.to_csv('/content/drive/MyDrive/BigMess/NASA/monthly_avg/3years/NASA_monthly_avg_2022.csv')
+
+# 2020
+nasa_2020_monthly_avg_3 = result_df_3.filter(f'Year = {2020}')
+pd_2020_monthly_avg_3 = nasa_2020_monthly_avg_3.toPandas()
+
+pd_2020_monthly_avg_3.to_csv('/content/drive/MyDrive/BigMess/NASA/monthly_avg/3years/NASA_monthly_avg_2020.csv')
+
+# 2000
+nasa_2000_monthly_avg_3 = result_df_3.filter(f'Year = {2000}')
+pd_2000_monthly_avg_3 = nasa_2000_monthly_avg_3.toPandas()
+
+pd_2000_monthly_avg_3.to_csv('/content/drive/MyDrive/BigMess/NASA/monthly_avg/3years/NASA_monthly_avg_2000.csv')
+
+# 1979
+nasa_1979_monthly_avg_3 = result_df_3.filter(f'Year = {1979}')
+pd_1979_monthly_avg_3 = nasa_1979_monthly_avg_3.toPandas()
+
+pd_1979_monthly_avg_3.to_csv('/content/drive/MyDrive/BigMess/NASA/monthly_avg/3years/NASA_monthly_avg_1979.csv')
```

Comment:
Skoro funkcja overYearMonthStats liczy średnią z danych z danego roku i n-lat wstecz, to jaki jest cel użycia tej funkcji dla ostatniego roku, dla którego mamy dane?

Reply:
Bardzo dobra uwaga.